### PR TITLE
[Benchmark] Benchmark OpCode and VM

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This pr creates a benchmark engine for the opcode and vm, making it easier to evaluate the opcodes and vm methods. 

## Type of change

- [ ] Optimization (the change is only an optimization)
- [x] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jim8y** [comment]
this pr is actually not complete, too many benchmarks need to be added, thus this pr should only focus on the benchmark framework.

File: benchmarks/Neo.VM.Benchmarks/InstructionBuilder/Instruction.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Numerics;
+using System.Reflection;
+using System.Text;
+
+namespace Neo.VM.Benchmark;
+
+[DebuggerDisplay("{_opCode}")]
+public class Instruction
```
Comment (shargon) [inline benchmarks/Neo.VM.Benchmarks/InstructionBuilder/Instruction.cs:24]: Why we need to replicate this class?

Comment (Jim8y) [reply benchmarks/Neo.VM.Benchmarks/InstructionBuilder/Instruction.cs:24]: this is from the compiler, it is different from what VM has, totally different.

Comment (shargon) [reply benchmarks/Neo.VM.Benchmarks/InstructionBuilder/Instruction.cs:24]: can we reference the other?

Comment (Jim8y) [reply benchmarks/Neo.VM.Benchmarks/InstructionBuilder/Instruction.cs:24]: > Can we use the current one?

we can not, what i want here is to constrcut the script, which is exactly what compiler instruction does. VM instruction is for execution purpose, they are totally different.

**Review by Hecate2 — APPROVED** [review]

**Review by vncoelho — APPROVED** [review]

File: benchmarks/Neo.VM.Benchmarks/OpCode/BenchmarkEngine.cs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.Test.Types;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace Neo.VM.Benchmark.OpCode;
```
Comment (shargon) [inline benchmarks/Neo.VM.Benchmarks/OpCode/BenchmarkEngine.cs:16]: some files has block namespace, and others don't

File: benchmarks/Neo.VM.Benchmarks/OpCode/Arrays/OpCode.Pack.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+namespace Neo.VM.Benchmark.OpCode;
+
+public class OpCode_Pack : OpCodeBase
+{
+
```
Comment (shargon) [inline benchmarks/Neo.VM.Benchmarks/OpCode/Arrays/OpCode.Pack.cs:16]:
```suggestion
```

File: benchmarks/Neo.VM.Benchmarks/OpCode/OpCode.Nop.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+// modifications are permitted.
+
+using BenchmarkDotNet.Attributes;
+
+namespace Neo.VM.Benchmark.OpCode;
+
+public class OpCode_Nop : OpCodeBase
+{
+    [GlobalSetup]
+    public void Setup()
```
Comment (shargon) [inline benchmarks/Neo.VM.Benchmarks/OpCode/OpCode.Nop.cs:19]: You want to call the base Setup? or override?

Comment (Jim8y) [reply benchmarks/Neo.VM.Benchmarks/OpCode/OpCode.Nop.cs]: I have deleted unfinished benchmark, this pr focus on creating the benchmark framework itself.

File: benchmarks/Neo.VM.Benchmarks/OpCode/BenchmarkEngine.cs (hunk 2 above)
Comment (Jim8y) [reply benchmarks/Neo.VM.Benchmarks/OpCode/BenchmarkEngine.cs:15]: updated.

**Pull request merged**